fendermint/vm/interpreter/src/chain.rs

```diff
@@ -89,32 +97,65 @@ where
+        .await?
+        {
+            None => {}
+            Some(proposal) => {
+                msgs.push(ChainMessage::Ipc(IpcMessage::TopDownExec(ParentFinality {
+                    height: proposal.height as ChainEpoch,
+                    block_hash: proposal.block_hash,
+                })))
+            }
+        }
+
         // Append at the end - if we run out of block space, these are going to be reproposed in the next block.
         msgs.extend(ckpts);
         Ok(msgs)
     }
 
     /// Perform finality checks on top-down transactions and availability checks on bottom-up transactions.
-    async fn process(&self, pool: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
+    async fn process(&self, state: Self::State, msgs: Vec<Self::Message>) -> anyhow::Result<bool> {
         for msg in msgs {
-            if let ChainMessage::Ipc(IpcMessage::BottomUpExec(msg)) = msg {
-                let item = CheckpointPoolItem::BottomUp(msg);
-
-                // We can just look in memory because when we start the application, we should retrieve any
-                // pending checkpoints (relayed but not executed) from the ledger, so they should be there.
-                // We don't have to validate the checkpoint here, because
-                // 1) we validated it when it was relayed, and
-                // 2) if a validator proposes something invalid, we can make them pay during execution.
-                let is_resolved = atomically(|| match pool.get_status(&item)? {
-                    None => Ok(false),
-                    Some(status) => status.is_resolved(),
-                })
-                .await;
-
-                if !is_resolved {
-                    return Ok(false);
+            match msg {
+                ChainMessage::Ipc(IpcMessage::BottomUpExec(msg)) => {
+                    let item = CheckpointPoolItem::BottomUp(msg);
+
+                    // We can just look in memory because when we start the application, we should retrieve any
+                    // pending checkpoints (relayed but not executed) from the ledger, so they should be there.
+                    // We don't have to validate the checkpoint here, because
+                    // 1) we validated it when it was relayed, and
+                    // 2) if a validator proposes something invalid, we can make them pay during execution.
+                    let is_resolved = atomically(|| match state.0.get_status(&item)? {
+                        None => Ok(false),
+                        Some(status) => status.is_resolved(),
+                    })
+                    .await;
+
+                    if !is_resolved {
+                        return Ok(false);
+                    }
                 }
-            }
+                ChainMessage::Ipc(IpcMessage::TopDownExec(ParentFinality {
+                    height,
+                    block_hash,
+                })) => {
+                    let prop = IPCParentFinality {
```

Comment:
Do we need both `ParentFinality` and `IPCParentFinality` to exist?

Reply:
I think it's just for crate completeness. Initially, `fendermint_vm_topdown` defined `IPCParentFinality` which should be imported into `fendermint_vm_message`. I think you mentioned once `fendermint_vm_message` should be kept thin and avoid import heavy crates such as `tokio` in `fendermint_vm_topdown`. But `fendermint_vm_topdown` import `message` seems strange to me. Since the `IPCParentFinality` is pretty light, so I just created a duplicate as a compromise.

Reply:
I have imported the `fendermint_vm_message` crate everywhere, but okay, if you prefer an almost identically named copy. Who knows it might have its own life.